sklearn/decomposition/_sparse_pca.py

```diff
@@ -29,6 +29,51 @@ def _check_normalize_components(normalize_components, estimator_name):
             )
 
 
+def _get_explained_variance(X, components, ridge_alpha):
+    """Get the explained variance.
+
+    Get the explained variance from the principal components of the
+    data. This follows the method outlined in [1] section 3.4 (Adjusted Total
+    Variance). For an alternate approach (not implemented here), see [2].
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_samples, n_features)
+        The feature vector. n_samples and n_features are the number of
+        samples and features, respectively.
+
+    components : array, shape (n_components, n_features)
+        The principal components. [1]
+
+    Notes
+    -----
+    Orthogonality is enforced in this case. Other variants exist that don't
+    enforce this [2].
+
+    References
+    ----------
+    .. [1] Hui Zou, Trevor Hastie, Robert Tibshirani ; Journal of Computational
+        and Graphical Statistics, Volume 15, Number 2, Pages 265–286.
+        DOI: 10.1198/106186006X113430
+    .. [2] Rodolphe Jenatton, Guillaume Obozinski, Francis Bach ; Proceedings
+        of the Thirteenth International Conference on Artificial Intelligence
+        and Statistics (AISTATS), PMLR 9:366-373, 2010.
+    """
+    # Transform input
+    t_spca = ridge_regression(components.T, X.T, ridge_alpha,
+                              solver="cholesky")
+
+    # QR decomposition of modified PCs
+    _, r = np.linalg.qr(t_spca)
+    variance = np.square(np.diag(r)) / (X.shape[0]-1)
+
+    # Variance in the original dataset
+    cov_mat = np.cov(X.T)
+    total_variance_in_x = np.trace(cov_mat) if cov_mat.size >= 2 \
+        else cov_mat
```

Comment:
how is it possible that cov_mat is not 2d here?

Reply:
It happens in a test, I will give it a closer look

Reply:
Ok, the failing tests are `test_estimators[SparsePCA()-check_fit2d_1feature]` and `test_estimators[MiniBatchPCA()-check_fit2d_1feature]`

With only one feature the covariance matrix is in fact a scalar:
```python
>>> np.cov([[1,1]])
array(0.)
>>> np.cov([[1, 1], [1, 1]])
array([[0., 0.],
       [0., 0.]])
```